### PR TITLE
Refactor exposure check to class, clean up

### PR DESCRIPTION
**Pull request by joshgummersall**
This PR revamps exposure checking to support both Location and BlueTooth token matching.

For token matching works by just seeing if any UUIDs in downloaded CSVs are present in the local DB.

- Additionally, for each exposed Beacon, we find all the locations that overlap within a given duration and mark those exposed.
- For each exposed token found, we store the location id of the location that is closest to the midpoint of the Beacon time range.

Additional changes:
- Clean up unused methods
- Encapsulate CSV serialization in the `LocationModel` and `BeaconModel` classes.

### CODE REVIEW

File: lib/exposure/exposure.dart [hunk 1]
```diff
@@ -0,0 +1,5 @@
+abstract class Exposure<T> {
+  final List<T> items;
+  const Exposure(this.items);
+  Future<List<T>> getExposures(String data);
```
Comment (wescarr) [inline lib/exposure/exposure.dart:4]: Might just be nice to add some comments about what subclasses should be implementing and what the structure of the `data` looks like

File: lib/exposure/location.dart [hunk 2]
```diff
@@ -0,0 +1,49 @@
+import 'package:covidtrace/exposure/exposure.dart';
+import 'package:covidtrace/helper/datetime.dart';
+import 'package:covidtrace/storage/location.dart';
```
Comment (wescarr) [inline lib/exposure/location.dart:3]: I've just been using relative paths for files in our package. Seems like vs code's behavior is inconsistent when auto importing.

File: lib/exposure/location.dart [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+      : super(locations) {
+    locations.forEach((location) {
+      var timestamp = ceilUnixSeconds(location.timestamp, 60);
+      var cellID = location.cellID.parent(level).toToken();
+      if (_lookup[cellID] == null) {
+        _lookup[cellID] = new Map();
+      }
+      if (_lookup[cellID][timestamp] == null) {
+        _lookup[cellID][timestamp] = [];
+      }
```
Comment (wescarr) [inline lib/exposure/location.dart:19]:
```dart
_lookup[cellID] ??= new Map();
_lookup[cellID][timestamp] ??= []
```

File: lib/exposure/location.dart [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  }
+
+  @override
+  Future<List<LocationModel>> getExposures(String data) async {
+    var rows =
+        CsvToListConverter(shouldParseNumbers: false, eol: '\n').convert(data);
+
+    var exposures = List<LocationModel>();
+
+    await Future.forEach(rows, (row) async {
```
Comment (wescarr) [inline lib/exposure/location.dart:31]: There doesn't seem to be any actual async logic in here.

File: lib/exposure/location.dart (hunk 2 above)
Comment (joshgummersall) [reply lib/exposure/location.dart:3]: I think package scoped imports seem more common in other Dart projects - in fact it's nice so if a file that imports ten other project packages is moved to a subdirectory you don't have to update any of the import paths.

File: lib/helper/check_exposures.dart [hunk 5]
```diff
@@ -23,21 +25,26 @@ Future<bool> checkExposures() async {
         whereArgs: [
           threeWeeksAgo.toIso8601String()
         ]), // no use searching already exposed locations or older than 3 weeks
+    Future.value([] as List<BeaconUuid>)
```
[the remark this answers is not in the document]
Comment (joshgummersall) [reply lib/helper/check_exposures.dart:28]: Yes, just stubbed out for now.

File: lib/state.dart [hunk 6]
```diff
@@ -113,71 +114,99 @@ class AppState with ChangeNotifier {
     return success;
   }
 
-  Future<bool> sendReport(Map<String, dynamic> symptoms) async {
-    var success = false;
+  Future<bool> _submitSymptoms(
```
Comment (wescarr) [inline lib/state.dart:117]: @joshgummersall Just realized I had a bit of a duplicate effort here: #35

Comment (joshgummersall) [reply lib/state.dart:117]: Yeah, I'd say get that other PR merged up then you can take this one over.

**Comment by joshgummersall** [comment]
Can't approve my own PR but looks good to me.

**Comment by wescarr** [comment]
@joshgummersall Thanks for the review!

File: lib/helper/check_exposures.dart [hunk 7]
```diff
@@ -129,57 +101,114 @@ Future<bool> checkExposures() async {
...
+/// the provided duration window. This method sets the `location` property on each
+/// `BeaconModel` when a match is found. Additionally it returns the set of locations
+/// that matched against any of the provided beacons.
+///
+/// The default `window` Duration is 10 minutes.
+Future<Map<int, LocationModel>> matchBeaconsAndLocations(
+    Iterable<BeaconModel> beacons,
+    {Duration window}) async {
+  window ??= Duration(minutes: 10);
+  var start = beacons.first.start.subtract(window).toIso8601String();
```
Comment (joshgummersall) [inline lib/helper/check_exposures.dart:165]: Do these Beacons need to be sorted? Worth making that explicit I'd say.

File: lib/helper/check_exposures.dart [hunk 8]
```diff
@@ -129,57 +101,114 @@ Future<bool> checkExposures() async {
...
+
+  List<LocationModel> locations = await LocationModel.findAll(
+      orderBy: 'id ASC',
+      where:
+          'DATETIME(timestamp) >= DATETIME(?) AND DATETIME(timestamp) <= DATETIME(?)',
+      whereArgs: [start, end]);
+  Map<int, LocationModel> locationMatches = {};
+
+  /// For each Beacon do the following:
+  /// - Find all the locations that fall within the duration window if the Beacon time range.
```
Comment (joshgummersall) [inline lib/helper/check_exposures.dart:176]: *window of the Beacon

File: lib/helper/check_exposures.dart [hunk 9]
```diff
@@ -129,57 +101,114 @@ Future<bool> checkExposures() async {
...
+      return aDiff.compareTo(bDiff);
+    });
+
+    if (matches.isNotEmpty) {
+      b.location = matches.first;
+      matches.forEach((l) => locationMatches[l.id] = l);
+    }
+  });
+
+  return locationMatches;
```
Comment (joshgummersall) [inline lib/helper/check_exposures.dart:202]: This could possibly be a `Set`?

Comment (wescarr) [reply lib/helper/check_exposures.dart:202]: I'd rather keep it as a map since since it matches the type of `exposedLocations`

**Pull request merged**